mlflow_client/core.py

```diff
@@ -0,0 +1,235 @@
+    """
+    Prevent logging if no active run is found in the current backend
+    """
+    def wrapper(self, *args):
+        if self._run_started:
+            return method(self, *args)
+        else:
+            raise ValueError("Logging method called without any active run")
+    return wrapper
+
+
+def get_auto_backend():
+    """
+    Returns a backend automatically
+    LocalBackend if mlflow is not found
+    MLFLow Local if mlflow is found
+    MLFlow Server if mlflow is found and MLFLOW_TRACKING_URI defined in env variables
+    """
+    try:
+        import mlflow
+        if 'MLFLOW_TRACKING_URI' in os.environ:
+            return MLFlowBackend(os.environ['MLFLOW_TRACKING_URI'])
+        else:
+            return MLFlowBackend()
+    except ImportError:
+        return LocalBackend()
+
+
+class Backend(ABC):
+    @abstractmethod
+    def log_metric(self, metric: str, value: int):
+        """
+        Log a metric (only numeric)
+        """
+        pass
+
+    @abstractmethod
+    def log_parameter(self, parameter: str, value):
+        """
+        Log a parameter that was used to launch the train, value can be any type
+        """
+        pass
+
+    @abstractmethod
+    def log_artifact(self, path_to_file: str, path_to_save:str = None):
+        """
+        Save a file, can be text, image, bytes ...
+        :param path_to_file: the path to the local file that needs to be saved as an artifact
+        :param path_to_save: optional path, if None: the artifact will be stored under /artifacts
+        if specified it will be stored under /artifacts/path_to_save
+        """
+        pass
+
+    @abstractmethod
+    def log_model(self,  model, name: str, library: MLFrameworks = MLFrameworks.PYTHON):
+        """
+        Save the resulting model of the run
+        :param model: an instance of a class
+        :param name: the name of the model
+        :param library: the framework used to train the model, must be one of @MLFrameworks
+        :return:
+        """
+        pass
+
+    @abstractmethod
+    def start_run(self, run_id: int = None):
+        """
+        Start a new run
+        :param run_id: optional run id parameter, if filled, will be used to log into an existing run
+        """
+        pass
+
+    @abstractmethod
+    def end_run(self):
+        """
+        End the current run
+        """
+        pass
+
+
+class LocalBackend(Backend):
+    """
+    Store metrics, parameters and artifacts without any need of importing mlflow
+    """
+    def __init__(self, path=None):
+        root_path = path if path else os.path.dirname(__file__)
+        self._path = os.path.join(root_path, 'mlruns')
+        self._run_started = 0
+
+    @if_run_active
+    def log_metric(self, metric: str, value: int):
+        path = os.path.join(self._path, str(self._run_started), 'metrics', metric)
+        os.makedirs(os.path.dirname(path), exist_ok=True)
+        with open(path, 'w') as f:
+            f.write(str(value))
```

Comment:
log in JSON format, better if we build some tools to reaggregate all files of a local folder.

```
{ "name": metric, "value": value }
```

Reply:
I did {"metricName": [metricValues]} instead